### PR TITLE
[PF-886] Clean up test workspaces with Janitor

### PR DESCRIPTION
**Pull request by zloery**
This change modifies all unit tests which create real workspaces to attempt to register them with Janitor to ensure they're always cleaned up.  This is currently only configured for `broad-dev` which we use for PR and nightly tests - if anyone is using other personal/developer environments I can add those to the Janitor configuration later.

I also moved pubsub topic/projectID to test config, which we'll need later for PF-963.

### CODE REVIEW

File: src/test/java/harness/utils/WorkspaceUtils.java
Comment (melissachang) [inline src/test/java/harness/utils/WorkspaceUtils.java:18]: Drive-by review. I'm also creating a WorkspaceUtils in my branch. :)

Can this be `createWorkspace` instead of `createWorkspaceWithCleanup`?

Comment (zloery) [reply src/test/java/harness/utils/WorkspaceUtils.java]: honestly I'm a little surprised it didn't exist already. and sg, done

File: src/test/java/harness/CRLJanitor.java
Comment (melissachang) [inline src/test/java/harness/CRLJanitor.java:54]: Opt - Hmm, in your Janitor config map, should we switch from `dev` to `broad-dev`, so we don't have to maintain this map?

Or, instead of maintaining 2 maps:
- In Janitor config, instead of map from WSM nickname to WSM url, just have a list of WSM urls
- In test config, put WSM url

File: src/test/java/harness/TestConfig.java
Comment (melissachang) [inline src/test/java/harness/TestConfig.java:25]: Now that Janitor cleans up workspaces, can you rename this to `useJanitor` ?

File: src/test/java/harness/TestConfig.java
Comment (melissachang) [inline src/test/java/harness/TestConfig.java:28]: Optional - This is the same for all envs right? Can this be a constant instead of test config?

File: src/test/java/harness/utils/WorkspaceUtils.java
Comment (melissachang) [inline src/test/java/harness/utils/WorkspaceUtils.java:31]: opt My personal preference is to omit these `@param` javadocs. go/java-style says `Javadoc is optional for "simple, obvious" members`. Something like

```
* @param name The name of the workspace
```

isn't adding anything.

File: src/test/java/harness/CRLJanitor.java
Comment (melissachang) [inline src/test/java/harness/CRLJanitor.java:114]: You have 3 methods:

1. registerWorkspaceForCleanupIfEnabled
2. registerWorkspaceForCleanup
3. registerWorkspaceForCleanup takes in instanceId

Callers should only call 1, right? Can these three combined into method?

```
  public static void registerWorkspaceForCleanup(
      UFWorkspace workspace, TestUser testUser) {
    if (!TestConfig.get().getUseJanitorForExternalResourcesCreatedByTests()) {
      return;
    }
    ...
```

3 throws an exception if `UseJanitorForExternalResourcesCreatedByTests` isn't true. I don't think we should throw an exception if `UseJanitorForExternalResourcesCreatedByTests` isn't true. We should be able to turn off `UseJanitorForExternalResourcesCreatedByTests` indepedent of code.

File: src/test/java/harness/utils/WorkspaceUtils.java
Comment (melissachang) [inline src/test/java/harness/utils/WorkspaceUtils.java:20]: // `terra workspace create`

File: src/test/java/harness/utils/WorkspaceUtils.java
Comment (melissachang) [inline src/test/java/harness/utils/WorkspaceUtils.java:37]: Take out `--format=json`

File: src/test/java/harness/CRLJanitor.java
Comment (zloery) [reply src/test/java/harness/CRLJanitor.java:54]: Hm, good point. It's kind of clunky keeping a bunch of hardcoded strings in sync across multiple repos, keeping a url allowlist in Janitor config would be cleaner but would still prevent calls to arbitrary URLs.
opened https://github.com/DataBiosphere/terra-resource-janitor/pull/140 for this, I'll update this PR after that one's merged

File: src/test/java/harness/TestConfig.java
Comment (zloery) [reply src/test/java/harness/TestConfig.java:28]: It's the same for all Broad environments but would be different for a Verily deployment, so I think this needs to live in TestConfig

File: src/test/java/harness/utils/WorkspaceUtils.java
Comment (zloery) [reply src/test/java/harness/utils/WorkspaceUtils.java]: Good point. I think it's worth documenting that `workspaceCreator` will be impersonated in a WSM call here, but the other params aren't adding anything.

File: src/test/java/harness/CRLJanitor.java
Comment (zloery) [reply src/test/java/harness/CRLJanitor.java]: Good point. I wrote all three thinking they would be useful, but after changing the tests I think this only needs the one method.

File: src/test/java/harness/utils/WorkspaceUtils.java
Comment (zloery) [reply src/test/java/harness/utils/WorkspaceUtils.java:21]: `runAndParseCommandExpectSuccess` always adds the "--format=json" argument, this is actually running // `terra workspace create --format=json`

File: src/test/java/harness/CRLJanitor.java
Comment (melissachang) [inline src/test/java/harness/CRLJanitor.java:84]: PubSub

File: src/test/java/harness/TestConfig.java
Comment (melissachang) [inline src/test/java/harness/TestConfig.java:95]: useJanitor

File: src/test/java/harness/TestConfig.java
Comment (melissachang) [inline src/test/java/harness/TestConfig.java:23]: Update comment

**Review by melissachang — APPROVED** [review]
Thanks! Upcoming changes are straightforward, so approving now

File: src/test/java/harness/CRLJanitor.java
Comment (jaycarlton) [inline src/test/java/harness/CRLJanitor.java:131]: so `get()` waits for completion?

File: src/test/java/harness/baseclasses/SingleWorkspaceUnit.java
Comment (jaycarlton) [inline src/test/java/harness/baseclasses/SingleWorkspaceUnit.java:33]: nit: `createdWorkspace`

File: src/test/java/harness/utils/WorkspaceUtils.java
Comment (jaycarlton) [inline src/test/java/harness/utils/WorkspaceUtils.java:35]: maybe name these `createTrackedWorkspace` so folks know they're special. I wouldn't have associated these calls with Janitor since the class is just `WorkspaceUtils`.

File: src/test/java/unit/CloneWorkspace.java
Comment (jaycarlton) [inline src/test/java/unit/CloneWorkspace.java:112]: for clone, the destination workspace won't be tracked automatically. We'll have to make sure the tests clean that up manually.

File: src/test/resources/testconfigs/broad.json
Comment (jaycarlton) [inline src/test/resources/testconfigs/broad.json:30]: it's kinda redundant to have `-pubsub-topic` in a topic name, but I guess this exists already.

Also, is this config for all Broad envs?

**Review by jaycarlton — APPROVED** [review]

File: src/test/java/harness/utils/WorkspaceUtils.java
Comment (melissachang) [reply src/test/java/harness/utils/WorkspaceUtils.java:45]: (Given that `runAndParseCommandExpectSuccess` and this method return `UFWorkspace`, I think it's ok to use here)

File: src/test/java/harness/CRLJanitor.java
Comment (zloery) [reply src/test/java/harness/CRLJanitor.java:127]: yep!

File: src/test/java/harness/utils/WorkspaceUtils.java
Comment (zloery) [reply src/test/java/harness/utils/WorkspaceUtils.java:46]: Most tests run commands the same way users do (as full commands) rather than calling internal methods, so I think UF object is right here.

(Whether we really should be doing that in unit tests is a whole separate discussion, too...)

File: src/test/java/harness/utils/WorkspaceUtils.java
Comment (zloery) [reply src/test/java/harness/utils/WorkspaceUtils.java:36]: Both names seem reasonable but I'd lean towards keeping this short - from a test reader perspective it's not important that the test objects are cleaned up later, it's the `createWorkspace` behavior I care about.

Added a comment about leaking workspaces if tests don't use this method

File: src/test/java/unit/CloneWorkspace.java
Comment (zloery) [reply src/test/java/unit/CloneWorkspace.java:112]: Yeah, definitely. Those tests can call `CRLJanitor.registerWorkspaceForCleanup`, but we'll have to be sure we're actually doing that

File: src/test/resources/testconfigs/broad.json
Comment (zloery) [reply src/test/resources/testconfigs/broad.json:30]: agreed, but the topic already exists so 🤷 
yes, we plan to use the same Janitor instance (and same pubsub topic) to clean up after all envs, though that's currently just dev.

**Pull request merged**
